### PR TITLE
Bst

### PR DESCRIPTION
**Pull request by Jakeand3rson**

### CODE REVIEW

File: bst.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    '''Instantiate Node and add helper functions'''
+    def __init__(self, value, parent=None, left_child=None, right_child=None):
+        self.value = value
+        self.parent = parent
+        self.left = left_child
+        self.right = right_child
+
+    def is_root(self):
+        '''Helper function for root node'''
+        return not self.parent
```
Comment (cewing) [inline bst.py:19]: good idea creating these helpers.  They should come in handy

File: bst.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+    def is_leaf(self):
+        ''''Helper function for acknowledging leaf'''
+        return not (self.right_child or self.left_child)
+
+
+class BST(object):
+    '''Instantiate binary search tree'''
+    def __init__(self, values=None):
```
Comment (cewing) [inline bst.py:47]: What would happen if the user passed values into this init function?  Anything?  Why is the argument there?

File: bst.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+                    break
+            elif current_node.value < value:
+                if current_node.right:
+                    current_node = current_node.right
+                else:
+                    current_node.right = BSTNode(value)
+                    self._size += 1
+                    break
+            else:
+                break
```
Comment (cewing) [inline bst.py:78]: I'm a bit bothered by the amount of repetition in this method.  You have three places you might set self._size, three places you might create a new node, etc.  I wonder if there might not be a way to streamline the process.

File: bst.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+if __name__ == '__main__':
+
+    def easy_tree():
+        x = random.sample(range(100), 100)
+        bst = BST()
+        bst.insert(50)
+        bst.insert(42.1)
+        for i in x:
+            bst.insert(i)
+        bst.contains(42.1)
```
Comment (cewing) [inline bst.py:153]: What does this tree look like?  What does using a random sampling of data do?  And since 42.1 is the second value inserted into the tree, is it ever anything but the first item to the left of the root node 50?

File: bst.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+    def hard_tree():
+        x = random.sample(range(100), 100)
+        bst = BST()
+        bst.insert(50)
+        for i in x:
+            bst.insert(i)
+        bst.contains(x[50])
```
Comment (cewing) [inline bst.py]: 50 is always the root, right?  Does this demonstrate anything since the first node examined will always be the root?

File: test_bst.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    for i in range(10):
+        t.insert(i)
+    assert t.depth() == 10
+
+
+def test_random():
+    t = BST()
+    for i in random.sample(range(100), 100):
+        t.insert(i)
+    assert t.size() == 100
```
Comment (cewing) [inline test_bst.py:66]: I would greatly enjoy seeing you test something other than a presorted data set or a random data set.  Building trees with well-understood structures by passing in specific numbers so that you can test edge cases of your algorithm is pretty important.  You are missing a lot of possible errors in these tests.

**Comment by Jakeand3rson** [comment]
Adjustments have been made and dot graph is working meow

File: bst.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+    t.communicate(dot_graph)
+
+    def easy_tree():
+        x = random.sample(range(100), 100)
+        bst = BST()
+        bst.insert(50)
+        for i in x:
+            bst.insert(i)
+        bst.insert(42.1)
+        bst.contains(42.1)
```
Comment (cewing) [inline bst.py:153]: better, but this is not the 'easy' situation.  It's sort of the "average" case.  The easy case would be a perfectly formed tree where the middle value of your data set was the root and every node was inserted in the "right" order.  See if you can make a method that does that.
